### DISCUSSION TITLE
new Plugin api `after` option

### DISCUSSION THREAD

**Discussion by Mister-Hope** [vuepress/core#1240] [Ideas]
Since plugins may access other plugins data, for example, `plugin-git` generate createTIme, updateTime and contributors, and these can be used in:

- Blogging
- SEO
- Feed
- ...

So related plugins may would like to have `plugin-git` being called first.

However, the squence caliing plugin can be problematic.

E.g.: For default theme. git plugin is called inside theme, and if users choose to call blogPlugin manually. blogPlugin will execlute first in every lifecycle, so it will fail to read git information. Another case can be `vuepress-theme-hope`, where the both plugins are built-in.

So to let everything work well, as long as users are calling any plugin, he shall call all plugins which the plugin is related to first manually. I.E., for the case above, users must call gitPlugin manually before blogPlugin to let the second one read git information.

---

## New API `after`

markdown-it plugins gives me insperations. If a plugin requires it running after another one, we can introduce a `after` field for it to declare:

e.g.: for blog plugin, it may add `after: ['@vuepress/plugin-git']`, and for feed plugin, it may add `after: ['@vuepress/plugin-git', 'vuepress-plugin-seo2']`

For our side, we shall read `after` option and sort the final plugins array according to it, so that every plugin can read other plugin data if needed.

User/theme plugin priority and plugin can not call plugins behavior can be all remain unchanged. This is a greate way to provide "Soft require support" between plugins.

@meteorlxy

**Comment by Mister-Hope** [comment]
Any feedback with this api proposal? @meteorlxy

**Comment by DreamOfIce** [comment]
Maybe there could also be a `before` option?

**Reply by Mister-Hope** [reply]
The before option is not needed because users plugin are called first.(higher priority

**Reply by DreamOfIce** [reply]
> The before option is not needed because users plugin are called first.(higher priority

But when both plugins are added by the user/theme, we have to tell them in the documentation that they have to add one or the other first. It might be better if this could be set within the plugin

**Reply by DreamOfIce** [reply]
> The before option is not needed because users plugin are called first.(higher priority

Also if the problem is reversed. Plugins that need to be loaded first are added by the theme, plugins that need to be loaded later may be added by the user

**Reply by Mister-Hope** [reply]
> Also if the problem is reversed,

I personally think you want to use the word "preserved"

**Reply by Mister-Hope** [reply]
I explain carefully in my discussion.

- If a plugin is not having `multiple: true`, then only the first instance is handled, and the rest are dropped.
- User plugins have the highest priority; we did this by adding user plugin first.
- There should be no problem within user's plugin or theme plugins, as theme developers and users are expected to provide a correct sequence.
- The problem only happens where the user calls a theme plugin, or user calls a plugin which should be run after some theme plugins. In this scenario, the `after` plugin can take effect, any plugin which relies on data coming from other plugins will be run after them correctly.
- If you want to pre-process some actions before the whole plugin/theme api system, then creating a local plugin and making it as the first user plugin will be the correct choice. If you want to insert your plugin between 2 theme plugins, declaring `after` with the first one solves this either.

---

Use blog2 plugin as an example

- When it's called by yourself, then you can add plugins before and after it perfectly in user plugin option.
- When it's called by theme,
    - Any user plugin is called first, so the pre process is working fine
    - If you want some plugins running after blog plugin, declare `after: ['vuepress-plugin-blog2']` in your plugin solves this.


List one if you still think there are any edge case which `after` do not solve your problem.

**Reply by DreamOfIce** [reply]
> > Also if the problem is reversed,
> 
> I personally think you want to use the word "preserved"

There's nothing wrong with the word. I did not express myself clearly, it has been updated 😅

**Comment by DreamOfIce** [comment]
实际运行似乎先加载的主题插件，而且源码中也是这样的：
https://github.com/vuepress/vuepress-next/blob/5d60575a9486195d45f4349324247fe2c6fcc090/packages/core/src/app/setupAppThemeAndPlugins.ts#L15-L18

**Reply by DreamOfIce** [reply]
@Mister-Hope @meteorlxy
